nilearn/mass_univariate/_utils.py

```diff
@@ -0,0 +1,81 @@
+"""Utility functions for the permuted least squares method."""
+import nibabel as nib
+import numpy as np
+from scipy import ndimage
+from nilearn.masking import apply_mask, unmask
+
+
+def _calculate_tfce(scores_array, masker, E=0.5, H=2, dh=0.1):
+    """Calculate threshold-free cluster enhancement values for scores maps.
+
+    The TFCE calculation is implemented as described in [1]_.
+
+    Parameters
+    ----------
+    scores_array : :obj:`numpy.ndarray`, shape=(n_regressors, n_descriptors)
+        Scores (t-statistics) for a set of regressors.
+    masker
+    E : :obj:`float`, optional
+        Extent weight. Default is 0.5.
+    H : :obj:`float`, optional
+        Height weight. Default is 2.
+    dh : :obj:`float`, optional
+        Step size for TFCE calculation. Default is 0.1.
+
+    Returns
+    -------
+    tfce_arr : :obj:`numpy.ndarray`, shape=(n_regressors, n_descriptors)
+        TFCE values.
+
+    References
+    ----------
+    .. [1] Smith, S. M., & Nichols, T. E. (2009).
+       Threshold-free cluster enhancement: addressing problems of smoothing,
+       threshold dependence and localisation in cluster inference.
+       Neuroimage, 44(1), 83-98.
+    """
+    # Define connectivity matrix for cluster labeling
+    conn = ndimage.generate_binary_structure(3, 1)
+
+    scores_4d_img = unmask(scores_array.T, masker.mask_img_)
+    scores_4d = scores_4d_img.get_fdata()
+    tfce_4d = np.zeros_like(scores_4d)
+
+    for i_regressor in range(scores_4d.shape[3]):
+        scores_3d = scores_4d[..., i_regressor]
+
+        # Get the step right before the maximum z-statistic in the map
+        max_z = np.floor(np.max(scores_3d) / dh) * dh
+
+        for score_thresh in np.arange(dh, max_z + dh, dh):
+            # Threshold map
+            thresh_scores_3d = scores_3d.copy()
+            thresh_scores_3d[thresh_scores_3d < score_thresh] = 0
+
+            # Derive clusters
+            labeled_arr3d, n_clusters = ndimage.measurements.label(
+                thresh_scores_3d,
+                conn,
+            )
+
+            # Label each cluster with its extent
+            cluster_map = np.zeros(scores_3d.shape, int)
+            for cluster_val in range(1, n_clusters + 1):
+                bool_map = labeled_arr3d == cluster_val
+                cluster_map[bool_map] = np.sum(bool_map)
+
+            # Calculate each voxel's tfce value based on its cluster extent
+            # and z-value
+            tfce_step_values = (cluster_map**E) * (score_thresh**H)
```

Comment:
```suggestion
            tfce_step_values = (cluster_map ** E) * (score_thresh ** H)
```

Reply:
I can make this change, but I originally didn't have spaces around the power operators because `black` removes them automatically. I've been largely sticking with `black`'s preferred style, except for using single quotes per `nilearn`'s style guidelines. Would you still prefer me to add whitespace?

Reply:
I thought it was the PEP rule,  but this is clearly not essential.